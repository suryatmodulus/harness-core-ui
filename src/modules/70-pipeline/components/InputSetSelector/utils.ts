import type { IconName, SelectOption } from '@harness/uicore'
import type { EntityGitDetails, InputSetSummaryResponse } from 'services/pipeline-ng'
import css from './InputSetSelector.module.scss'

type InputSetLocal = InputSetSummaryResponse & SelectOption
export interface InputSetValue extends InputSetLocal {
  type: InputSetSummaryResponse['inputSetType']
  gitDetails?: EntityGitDetails
}

export const getIconByType = (type: InputSetSummaryResponse['inputSetType']): IconName => {
  return type === 'OVERLAY_INPUT_SET' ? 'step-group' : 'yaml-builder-input-sets'
}

export const onDragStart = (event: React.DragEvent<HTMLLIElement>, row: InputSetValue): void => {
  event.dataTransfer.setData('data', JSON.stringify(row))
  event.currentTarget.classList.add(css.dragging)
}

export const onDragEnd = (event: React.DragEvent<HTMLLIElement>): void => {
  event.currentTarget.classList.remove(css.dragging)
}

export const onDragLeave = (event: React.DragEvent<HTMLLIElement>): void => {
  event.currentTarget.classList.remove(css.dragOver)
}

export const onDragOver = (event: React.DragEvent<HTMLLIElement>): void => {
  if (event.preventDefault) {
    event.preventDefault()
  }
  event.currentTarget.classList.add(css.dragOver)
  event.dataTransfer.dropEffect = 'move'
}
