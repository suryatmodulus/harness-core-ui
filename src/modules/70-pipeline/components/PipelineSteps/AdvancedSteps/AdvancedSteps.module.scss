/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

.form {
  display: flex;
  flex-direction: column;
  [data-testid='delegateSelectors-panel'] :global {
    .bp3-form-group {
      margin-bottom: 0px;
    }
    .bp3-collapse {
      z-index: 2;
    }
  }
}

.skipConditionLabel {
  // TODO: Replace color once it is available in UIKit
  color: #223147 !important;
}
