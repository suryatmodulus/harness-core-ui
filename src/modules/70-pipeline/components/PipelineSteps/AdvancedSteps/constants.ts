/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

export const cvDefaultFailureStrategies = [
  {
    onFailure: {
      errors: ['Verification'],
      action: {
        type: 'ManualIntervention',
        spec: {
          timeout: '2h',
          onTimeout: {
            action: {
              type: 'StageRollback'
            }
          }
        }
      }
    }
  },
  {
    onFailure: {
      errors: ['AnyOther'],
      action: {
        type: 'ManualIntervention',
        spec: {
          timeout: '2h',
          onTimeout: {
            action: {
              type: 'Ignore'
            }
          }
        }
      }
    }
  }
]
