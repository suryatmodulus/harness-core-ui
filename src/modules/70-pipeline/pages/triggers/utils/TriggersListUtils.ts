/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import type { IconName } from '@wings-software/uicore'
// temporary mock data
import { parse } from 'yaml'
import type { AddDrawerMapInterface } from '@common/components/AddDrawer/AddDrawer'
import type { StringKeys } from 'framework/strings'
import { manifestTypeIcons, ManifestDataType } from '@pipeline/components/ManifestSelection/Manifesthelper'
import { ArtifactIconByType, ENABLED_ARTIFACT_TYPES } from '@pipeline/components/ArtifactsSelection/ArtifactHelper'
import { TriggerTypes, AWS_CODECOMMIT, AwsCodeCommit } from './TriggersWizardPageUtils'

export const GitSourceProviders: Record<string, { value: string; iconName: IconName }> = {
  GITHUB: { value: 'Github', iconName: 'github' },
  GITLAB: { value: 'Gitlab', iconName: 'service-gotlab' },
  BITBUCKET: { value: 'Bitbucket', iconName: 'bitbucket-selected' },
  AWS_CODECOMMIT: { value: 'AwsCodeCommit', iconName: 'service-aws-code-deploy' },
  CUSTOM: { value: 'Custom', iconName: 'build' }
}

const TriggerTypeIcons = {
  SCHEDULE: 'trigger-schedule',
  NEW_ARTIFACT: 'new-artifact'
}
export const getTriggerIcon = ({
  type,
  webhookSourceRepo,
  buildType
}: {
  type: string
  webhookSourceRepo?: string // string temporary until backend
  buildType?: string
}): IconName => {
  const updatedWebhookSourceRepo =
    webhookSourceRepo === AwsCodeCommit ? AWS_CODECOMMIT : webhookSourceRepo?.toUpperCase()
  const webhookSourceRepoIconName =
    webhookSourceRepo && updatedWebhookSourceRepo && GitSourceProviders[updatedWebhookSourceRepo]?.iconName
  if (type === TriggerTypes.WEBHOOK && webhookSourceRepoIconName) {
    return webhookSourceRepoIconName as IconName
  } else if (type === TriggerTypes.SCHEDULE) {
    return TriggerTypeIcons.SCHEDULE as IconName
  } else if (type === TriggerTypes.MANIFEST && buildType) {
    if (buildType === ManifestDataType.HelmChart) {
      return manifestTypeIcons.HelmChart
    }
  } else if (type === TriggerTypes.ARTIFACT && buildType) {
    switch (buildType) {
      case ENABLED_ARTIFACT_TYPES.Gcr:
        return ArtifactIconByType.Gcr
      case ENABLED_ARTIFACT_TYPES.Ecr:
        return ArtifactIconByType.Ecr
      case ENABLED_ARTIFACT_TYPES.DockerRegistry:
        return ArtifactIconByType.DockerRegistry
    }
  }
  return 'yaml-builder-trigger'
}

const triggerDrawerMap = (getString: (key: StringKeys) => string): AddDrawerMapInterface => ({
  drawerLabel: getString('pipeline.triggers.triggersLabel'),
  showAllLabel: getString('pipeline.triggers.showAllTriggers'),
  categories: [
    {
      categoryLabel: getString('execution.triggerType.WEBHOOK'),
      categoryValue: 'Webhook',
      items: [
        {
          itemLabel: getString('common.repo_provider.githubLabel'),
          value: GitSourceProviders.GITHUB.value,
          iconName: GitSourceProviders.GITHUB.iconName
        },
        {
          itemLabel: getString('common.repo_provider.gitlabLabel'),
          value: GitSourceProviders.GITLAB.value,
          iconName: GitSourceProviders.GITLAB.iconName
        },
        {
          itemLabel: getString('common.repo_provider.bitbucketLabel'),
          value: GitSourceProviders.BITBUCKET.value,
          iconName: GitSourceProviders.BITBUCKET.iconName
        },
        {
          itemLabel: getString('common.repo_provider.awscodecommit'),
          value: GitSourceProviders.AWS_CODECOMMIT.value,
          iconName: GitSourceProviders.AWS_CODECOMMIT.iconName
        },
        {
          itemLabel: getString('common.repo_provider.customLabel'),
          value: GitSourceProviders.CUSTOM.value,
          iconName: GitSourceProviders.CUSTOM.iconName
        }
      ]
    },
    {
      categoryLabel: getString('pipeline.triggers.scheduledLabel'),
      categoryValue: 'Scheduled',
      items: [
        {
          itemLabel: getString('pipeline.triggers.cronLabel'),
          value: 'Cron',
          iconName: TriggerTypeIcons.SCHEDULE as IconName
        }
      ]
    }
  ]
})

export const getSourceRepoOptions = (getString: (str: StringKeys) => string): { label: string; value: string }[] => [
  { label: getString('common.repo_provider.githubLabel'), value: GitSourceProviders.GITHUB.value },
  { label: getString('common.repo_provider.gitlabLabel'), value: GitSourceProviders.GITLAB.value },
  { label: getString('common.repo_provider.bitbucketLabel'), value: GitSourceProviders.BITBUCKET.value },
  { label: getString('common.repo_provider.codecommit'), value: GitSourceProviders.AWS_CODECOMMIT.value },
  { label: getString('common.repo_provider.customLabel'), value: GitSourceProviders.CUSTOM.value }
]

export const getCategoryItems = (getString: (key: StringKeys) => string): AddDrawerMapInterface =>
  triggerDrawerMap(getString)

export interface ItemInterface {
  itemLabel: string
  iconName: IconName
  value: string
  visible?: boolean
  disabled?: boolean
  categoryValue?: string
}

export interface TriggerDataInterface {
  triggerType: string
  sourceRepo?: string
  manifestType?: string
  artifactType?: string
}

export const getEnabledStatusTriggerValues = ({
  data,
  enabled,
  getString
}: {
  data: any
  enabled: boolean
  getString: (key: StringKeys) => string
}): { values?: any; error?: string } => {
  try {
    const triggerResponseJson = parse(data?.yaml || '')
    triggerResponseJson.trigger.enabled = enabled
    return { values: triggerResponseJson.trigger }
  } catch (e) {
    return { error: getString('pipeline.triggers.cannotParseTriggersData') }
  }
}
const TriggerStatus = {
  FAILED: 'FAILED',
  UNKNOWN: 'UNKNOWN',
  ERROR: 'ERROR',
  TIMEOUT: 'TIMEOUT',
  UNAVAILABLE: 'UNAVAILABLE',
  SUCCESS: 'SUCCESS'
}

export const errorStatusList = [
  TriggerStatus.FAILED,
  TriggerStatus.UNKNOWN,
  TriggerStatus.ERROR,
  TriggerStatus.TIMEOUT,
  TriggerStatus.UNAVAILABLE
]
