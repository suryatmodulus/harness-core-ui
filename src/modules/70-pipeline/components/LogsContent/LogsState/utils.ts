/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import { escapeStringRegexp } from '@common/utils/StringUtils'

import type { State } from './types'

export function getRegexForSearch(str: string): RegExp {
  return new RegExp(escapeStringRegexp(str), 'gi')
}

export function getDefaultReducerState(props: Partial<Pick<State, 'selectedStage' | 'selectedStep'>> = {}): State {
  return {
    units: [],
    logKeys: [],
    selectedStage: props.selectedStage || '',
    selectedStep: props.selectedStep || '',
    dataMap: {},
    searchData: {
      text: '',
      currentIndex: 0,
      linesWithResults: []
    }
  }
}

export const logsCache = new Map<string, string>()
