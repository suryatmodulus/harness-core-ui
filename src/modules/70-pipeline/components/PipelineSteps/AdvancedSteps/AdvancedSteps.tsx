/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React from 'react'
import type { FormikProps } from 'formik'
import { Formik, FormikForm, Accordion, AccordionHandle } from '@wings-software/uicore'
import * as Yup from 'yup'
import { debounce, defaultTo, isEmpty } from 'lodash-es'

import { useStrings } from 'framework/strings'
import {
  AdvancedPanels,
  StepCommandsProps,
  Values,
  TabTypes
} from '@pipeline/components/PipelineStudio/StepCommands/StepCommandTypes'
import { StepFormikFowardRef, setFormikRef } from '@pipeline/components/AbstractSteps/Step'
import { StepMode as Modes } from '@pipeline/utils/stepUtils'
import type { StepElementConfig, StepGroupElementConfig } from 'services/cd-ng'
import type { TemplateStepNode } from 'services/pipeline-ng'
import DelegateSelectorPanel from './DelegateSelectorPanel/DelegateSelectorPanel'
import FailureStrategyPanel, { AllFailureStrategyConfig } from './FailureStrategyPanel/FailureStrategyPanel'
import { getFailureStrategiesValidationSchema } from './FailureStrategyPanel/validation'
import type { StepType } from '../PipelineStepInterface'
import ConditionalExecutionPanel from './ConditionalExecutionPanel/ConditionalExecutionPanel'
import css from './AdvancedSteps.module.scss'

export type FormValues = Pick<Values, 'delegateSelectors' | 'when'> & {
  failureStrategies: AllFailureStrategyConfig[]
}

export interface AdvancedStepsProps extends Omit<StepCommandsProps, 'onUseTemplate' | 'onRemoveTemplate'> {
  stepType?: StepType
}

export default function AdvancedSteps(props: AdvancedStepsProps, formikRef: StepFormikFowardRef): React.ReactElement {
  const { step, onChange, onUpdate } = props
  const { getString } = useStrings()

  const debouncedUpdate = React.useCallback(
    debounce((data: FormValues): void => {
      onChange?.({ ...data, tab: TabTypes.Advanced })
    }, 300),
    [onUpdate]
  )

  const failureStrategies =
    ((step as TemplateStepNode)?.template?.templateInputs as StepElementConfig)?.failureStrategies ||
    (step as StepElementConfig | StepGroupElementConfig)?.failureStrategies

  const delegateSelectors =
    ((step as TemplateStepNode)?.template?.templateInputs as StepElementConfig)?.spec?.delegateSelectors ||
    (step as StepElementConfig)?.spec?.delegateSelectors

  const when =
    ((step as TemplateStepNode)?.template?.templateInputs as StepElementConfig)?.when ||
    (step as StepElementConfig | StepGroupElementConfig)?.when

  return (
    <Formik<FormValues>
      initialValues={{
        failureStrategies: defaultTo(failureStrategies, []) as AllFailureStrategyConfig[],
        delegateSelectors: defaultTo(delegateSelectors, []),
        when
      }}
      onSubmit={data => {
        onUpdate({ ...data, tab: TabTypes.Advanced })
      }}
      validate={debouncedUpdate}
      formName="pipelineAdvancedSteps"
      validationSchema={Yup.object().shape({
        failureStrategies: getFailureStrategiesValidationSchema(getString)
      })}
    >
      {(formikProps: FormikProps<FormValues>) => {
        setFormikRef(formikRef, formikProps)

        return <AdvancedTabForm {...props} formikProps={formikProps} />
      }}
    </Formik>
  )
}

export interface AdvancedTabFormProps extends Omit<AdvancedStepsProps, 'onChange'> {
  formikProps: FormikProps<FormValues>
}

export function AdvancedTabForm(props: AdvancedTabFormProps): React.ReactElement {
  const {
    formikProps,
    hiddenPanels = [],
    hasStepGroupAncestor,
    isStepGroup,
    stepsFactory,
    isReadonly,
    stageType,
    stepType
  } = props

  const accordionRef = React.useRef<AccordionHandle>({} as AccordionHandle)
  const { getString } = useStrings()

  React.useEffect(() => {
    if (formikProps.isSubmitting) {
      if (!isEmpty(formikProps.errors?.failureStrategies) && accordionRef.current) {
        accordionRef.current.open(AdvancedPanels.FailureStrategy)
      }

      if (!isEmpty(formikProps.errors?.when) && accordionRef.current) {
        accordionRef.current.open(AdvancedPanels.ConditionalExecution)
      }

      if (!isEmpty(formikProps.errors?.delegateSelectors) && accordionRef.current) {
        accordionRef.current.open(AdvancedPanels.DelegateSelectors)
      }
    }
  }, [formikProps.isSubmitting, formikProps.errors])

  return (
    <FormikForm className={css.form}>
      <div>
        <Accordion
          ref={accordionRef}
          allowMultiOpen
          activeId={
            hiddenPanels.indexOf(AdvancedPanels.DelegateSelectors) === -1 &&
            stepsFactory.getStep(stepType)?.hasDelegateSelectionVisible
              ? AdvancedPanels.DelegateSelectors
              : hiddenPanels.indexOf(AdvancedPanels.ConditionalExecution) === -1
              ? AdvancedPanels.ConditionalExecution
              : hiddenPanels.indexOf(AdvancedPanels.FailureStrategy) === -1
              ? AdvancedPanels.FailureStrategy
              : ''
          }
        >
          {hiddenPanels.indexOf(AdvancedPanels.DelegateSelectors) === -1 &&
            stepsFactory.getStep(stepType)?.hasDelegateSelectionVisible && (
              <Accordion.Panel
                id={AdvancedPanels.DelegateSelectors}
                summary={getString('delegate.DelegateSelector')}
                details={<DelegateSelectorPanel isReadonly={isReadonly} formikProps={formikProps} />}
              />
            )}
          {hiddenPanels.indexOf(AdvancedPanels.ConditionalExecution) === -1 && (
            <Accordion.Panel
              id={AdvancedPanels.ConditionalExecution}
              summary={getString('pipeline.conditionalExecution.title')}
              details={
                <ConditionalExecutionPanel
                  formikProps={formikProps}
                  mode={isStepGroup ? Modes.STEP_GROUP : Modes.STEP}
                  isReadonly={isReadonly}
                />
              }
            />
          )}
          {hiddenPanels.indexOf(AdvancedPanels.FailureStrategy) === -1 && (
            <Accordion.Panel
              id={AdvancedPanels.FailureStrategy}
              summary={getString('pipeline.failureStrategies.title')}
              details={
                <FailureStrategyPanel
                  mode={hasStepGroupAncestor || isStepGroup ? Modes.STEP_GROUP : Modes.STEP}
                  stageType={stageType}
                  formikProps={formikProps}
                  isReadonly={isReadonly}
                />
              }
            />
          )}
        </Accordion>
      </div>
    </FormikForm>
  )
}

export const AdvancedStepsWithRef = React.forwardRef(AdvancedSteps)
