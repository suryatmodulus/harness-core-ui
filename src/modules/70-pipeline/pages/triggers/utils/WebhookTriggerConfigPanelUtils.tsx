/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React, { SetStateAction, Dispatch } from 'react'
import cx from 'classnames'
import { FormInput, SelectOption, Text, Container, FontVariation, Color } from '@wings-software/uicore'
import { isEmpty, isUndefined } from 'lodash-es'
import type { StringKeys, UseStringsReturn } from 'framework/strings'
import { GitSourceProviders } from './TriggersListUtils'
import { ConnectorSection } from '../views/ConnectorSection'
import { eventTypes } from './TriggersWizardPageUtils'
import css from '../views/WebhookTriggerConfigPanel.module.scss'

export const autoAbortPreviousExecutionsTypes = [
  eventTypes.PUSH,
  eventTypes.PULL_REQUEST,
  eventTypes.ISSUE_COMMENT,
  eventTypes.MERGE_REQUEST
]

export const getAutoAbortDescription = ({
  event,
  getString
}: {
  event: string
  getString: (key: StringKeys) => string
}): string => {
  if (event === eventTypes.PUSH) {
    return getString('pipeline.triggers.triggerConfigurationPanel.autoAbortPush')
  } else if (event === eventTypes.PULL_REQUEST || event === eventTypes.MERGE_REQUEST) {
    return getString('pipeline.triggers.triggerConfigurationPanel.autoAbortPR')
  } else if (event === eventTypes.ISSUE_COMMENT) {
    return getString('pipeline.triggers.triggerConfigurationPanel.autoAbortIssueComment')
  }
  return ''
}

export const handleSourceRepoChange = ({ e, formikProps }: { e: SelectOption; formikProps: any }): void => {
  if (e.value === GitSourceProviders.CUSTOM.value) {
    formikProps.setValues({
      ...formikProps.values,
      sourceRepo: e.value,
      connectorRef: undefined,
      repoName: '',
      actions: undefined,
      anyAction: false,
      secretToken: undefined
    })
  } else {
    formikProps.setValues({
      ...formikProps.values,
      sourceRepo: e.value,
      connectorRef: undefined,
      repoName: '',
      actions: undefined,
      anyAction: false,
      secretToken: undefined,
      headerConditions: undefined
    })
  }
}

export const renderNonCustomEventFields = ({
  sourceRepo,
  formikProps,
  event,
  eventOptions,
  getString,
  actionsOptions,
  actions
}: {
  sourceRepo?: string
  formikProps: any
  event?: string
  eventOptions: SelectOption[]
  getString: UseStringsReturn['getString']
  actionsOptions: SelectOption[]
  actions: SelectOption[]
}): JSX.Element => {
  return (
    <>
      {sourceRepo && <ConnectorSection formikProps={formikProps} />}
      <FormInput.Select
        className={cx(!event && css.bottomMarginZero)}
        key={event}
        label={getString('pipeline.triggers.triggerConfigurationPanel.event')}
        name="event"
        placeholder={getString('pipeline.triggers.triggerConfigurationPanel.eventPlaceholder')}
        items={eventOptions}
        onChange={e => {
          const additionalValues: any = {}

          if (e.value === eventTypes.PUSH) {
            additionalValues.sourceBranchOperator = undefined
            additionalValues.sourceBranchValue = undefined
          }

          if (e.value !== eventTypes.TAG) {
            additionalValues.tagConditionOperator = undefined
            additionalValues.tagConditionValue = undefined
          }

          formikProps.setValues({
            ...formikProps.values,
            event: e.value,
            actions: e.value === eventTypes.PUSH ? [] : undefined,
            anyAction: false,
            ...additionalValues
          })
        }}
      />
      {event && (
        <>
          {actionsOptions?.length !== 0 && (
            <div className={css.actionsContainer}>
              <div>
                <Text
                  font={{ variation: FontVariation.FORM_INPUT_TEXT, weight: 'semi-bold' }}
                  color={Color.GREY_600}
                  style={{ marginBottom: 'var(--spacing-xsmall)' }}
                >
                  {getString('pipeline.triggers.triggerConfigurationPanel.actions')}
                </Text>
                <FormInput.MultiSelect
                  className={css.multiSelect}
                  name="actions"
                  items={actionsOptions}
                  // yaml design: empty array means selecting all
                  disabled={Array.isArray(actions) && isEmpty(actions)}
                  onChange={e => {
                    if (!e || (Array.isArray(e) && isEmpty(e))) {
                      formikProps.setFieldValue('actions', undefined)
                    } else {
                      formikProps.setFieldValue('actions', e)
                    }
                  }}
                />
              </div>
              <Container className={css.anyActionContainer}>
                <FormInput.CheckBox
                  name="anyAction"
                  key={Date.now()}
                  label={getString('pipeline.triggers.triggerConfigurationPanel.anyActions')}
                  defaultChecked={(Array.isArray(actions) && actions.length === 0) || false}
                  className={css.checkboxAlignment}
                  onClick={(e: React.FormEvent<HTMLInputElement>) => {
                    formikProps.setFieldTouched('actions', true)
                    if (e.currentTarget?.checked) {
                      formikProps.setFieldValue('actions', [])
                    } else {
                      formikProps.setFieldValue('actions', undefined)
                    }
                  }}
                />
              </Container>
            </div>
          )}
          {autoAbortPreviousExecutionsTypes.includes(event) && (
            <>
              <FormInput.CheckBox
                style={{ position: 'relative', left: '0' }}
                name="autoAbortPreviousExecutions"
                label="Auto-abort Previous Execution"
                className={css.checkboxAlignment}
              />
              <Text className={css.autoAbortDescription}>{getAutoAbortDescription({ event, getString })}</Text>
            </>
          )}
        </>
      )}
    </>
  )
}
interface ActionsOptionsMapInterface {
  [key: string]: string[]
}

export const getEventLabelMap = (event: string): string => {
  // add space between camelcase-separated words
  if (event === eventTypes.PULL_REQUEST) {
    return 'Pull Request'
  } else if (event === eventTypes.MERGE_REQUEST) {
    return 'Merge Request'
  } else if (event === eventTypes.ISSUE_COMMENT) {
    return 'Issue Comment'
  }
  return event
}

export const getEventAndActions = ({
  data,
  sourceRepo
}: {
  data: {
    [key: string]: {
      [key: string]: string[]
    }
  }
  sourceRepo: string
}): { eventOptions: SelectOption[]; actionsOptionsMap: ActionsOptionsMapInterface } => {
  const filteredData = data?.[sourceRepo] || {}
  const eventOptions = Object.keys(filteredData).map(event => ({
    label: getEventLabelMap(event),
    value: event
  }))
  return { eventOptions, actionsOptionsMap: filteredData }
}

export const clearEventsAndActions = ({
  newActionsOptionsMap,
  formikProps,
  setActionsOptions,
  event
}: {
  newActionsOptionsMap: ActionsOptionsMapInterface
  formikProps: any
  setActionsOptions: Dispatch<SetStateAction<SelectOption[]>>
  event: string
}): void => {
  if (isUndefined(newActionsOptionsMap[event])) {
    formikProps.setFieldValue('event', undefined)
    return
  }

  const newActionsOptions = newActionsOptionsMap[event]?.map((val: string) => ({ label: val, value: val }))
  setActionsOptions(newActionsOptions)
  if (newActionsOptions?.length === 0) {
    formikProps.setFieldValue('actions', [])
  }
}
