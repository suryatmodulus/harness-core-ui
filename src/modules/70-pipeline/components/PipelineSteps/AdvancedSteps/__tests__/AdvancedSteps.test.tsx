/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React from 'react'
import { render, act } from '@testing-library/react'

import { MultiTypeInputType } from '@wings-software/uicore'
import { TestWrapper } from '@common/utils/testUtils'
import type { StepFormikRef } from '@pipeline/components/AbstractSteps/Step'

import { AdvancedStepsWithRef } from '../AdvancedSteps'

jest.mock('@common/components/MonacoEditor/MonacoEditor')

const mockGetCallFunction = jest.fn()
jest.mock('services/portal', () => ({
  useGetDelegateSelectorsUpTheHierarchy: jest.fn().mockImplementation(args => {
    mockGetCallFunction(args)
    return []
  })
}))

describe('<AdvancedSteps /> tests', () => {
  test('snapshot test', () => {
    const { container } = render(
      <TestWrapper>
        <AdvancedStepsWithRef
          isStepGroup={false}
          isReadonly={false}
          step={{} as any}
          allowableTypes={[MultiTypeInputType.FIXED, MultiTypeInputType.RUNTIME, MultiTypeInputType.EXPRESSION]}
          stepsFactory={{ getStep: jest.fn(() => ({ hasDelegateSelectionVisible: true })) } as any}
          onUpdate={jest.fn()}
        />
      </TestWrapper>
    )
    expect(container).toMatchSnapshot()
  })

  // this test can be removed if we remove obSubmit handler
  test('submit works', async () => {
    const ref = React.createRef<StepFormikRef<unknown>>()
    const onSubmit = jest.fn()

    render(
      <TestWrapper>
        <AdvancedStepsWithRef
          isStepGroup={false}
          step={{} as any}
          isReadonly={false}
          allowableTypes={[MultiTypeInputType.FIXED, MultiTypeInputType.RUNTIME, MultiTypeInputType.EXPRESSION]}
          stepsFactory={{ getStep: jest.fn() } as any}
          ref={ref}
          onUpdate={onSubmit}
        />
      </TestWrapper>
    )

    await act(() => ref.current?.submitForm())

    expect(onSubmit).toHaveBeenCalledWith({
      delegateSelectors: [],
      failureStrategies: [],
      when: undefined,
      tab: 'ADVANCED'
    })
  })
})
