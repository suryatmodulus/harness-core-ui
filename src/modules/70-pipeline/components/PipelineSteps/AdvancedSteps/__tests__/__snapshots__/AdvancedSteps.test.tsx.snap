// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AdvancedSteps /> tests snapshot test 1`] = `
<div>
  <div
    class="OverlaySpinner--overlaySpinner"
  >
    <form
      class="FormikForm--main form"
    >
      <div>
        <div>
          <div
            class="Accordion--accordion"
          >
            <div
              class="Accordion--panel"
              data-open="true"
              data-testid="delegateSelectors-panel"
            >
              <div
                class="Accordion--summary"
                data-testid="delegateSelectors-summary"
              >
                <div
                  class="Accordion--label"
                >
                  delegate.DelegateSelector
                </div>
                <div
                  class="Accordion--chevron"
                />
              </div>
              <div
                class="bp3-collapse Accordion--collapse"
                style="height: auto; overflow-y: visible; transition: none;"
              >
                <div
                  aria-hidden="false"
                  class="bp3-collapse-body"
                  style="transform: translateY(0); transition: none;"
                >
                  <div
                    class="Accordion--details"
                    data-testid="delegateSelectors-details"
                  >
                    <div
                      class="bp3-form-group"
                      data-id="undefined-0"
                    >
                      <div
                        class="bp3-form-content"
                      >
                        <div
                          class="Layout--horizontal StyledProps--main MultiTypeInput--main"
                          style="flex-grow: 1;"
                        >
                          <div
                            class="wrapper wrapper"
                            data-name="DelegateSelectors"
                          >
                            <div
                              class="bp3-popover-wrapper delegateInput delegatePopover bp3-fill"
                            >
                              <div
                                class="bp3-popover-target"
                              >
                                <div
                                  class=""
                                >
                                  <div
                                    class="bp3-input bp3-tag-input bp3-fill bp3-multi-select"
                                  >
                                    <div
                                      class="bp3-tag-input-values"
                                    >
                                      <input
                                        class="bp3-input-ghost bp3-multi-select-tag-input-input"
                                        placeholder="delegate.Delegate_Selector_placeholder"
                                        value=""
                                      />
                                    </div>
                                  </div>
                                </div>
                              </div>
                            </div>
                          </div>
                          <span
                            class="bp3-popover-wrapper MultiTypeInput--wrapper"
                          >
                            <span
                              class="bp3-popover-target"
                            >
                              <button
                                class="MultiTypeInput--btn MultiTypeInput--FIXED"
                              >
                                <span
                                  class="bp3-icon StyledProps--main"
                                  data-icon="fixed-input"
                                >
                                  <svg
                                    height="12"
                                    viewBox="0 0 9 9"
                                    width="12"
                                    xmlns="http://www.w3.org/2000/svg"
                                  >
                                    <path
                                      d="M5.61 7.55c.62-.7.3-1.42.18-2.35a.32.32 0 01.1-.29l1.43-1.42a.76.76 0 01.28-.15c.39-.1 1 0 1.28-.29a.34.34 0 000-.51L6.46.12a.33.33 0 00-.52 0c-.28.29-.2.9-.28 1.3a.53.53 0 01-.13.23L4.09 3.12a.29.29 0 01-.28.09c-.89-.11-1.6-.49-2.27.11-.29.26-.29.43 0 .7l1.31 1.34.15.16-.15.15-2.7 2.69a.36.36 0 000 .54.36.36 0 00.54-.05l2.63-2.68.17-.17.17.16c.45.46.9.91 1.36 1.36.21.21.42.22.59.03z"
                                      fill="currentColor"
                                    />
                                  </svg>
                                </span>
                              </button>
                            </span>
                          </span>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="Accordion--panel"
              data-open="false"
              data-testid="conditionalExecution-panel"
            >
              <div
                class="Accordion--summary"
                data-testid="conditionalExecution-summary"
              >
                <div
                  class="Accordion--label"
                >
                  pipeline.conditionalExecution.title
                </div>
                <div
                  class="Accordion--chevron"
                />
              </div>
              <div
                class="bp3-collapse Accordion--collapse"
              >
                <div
                  aria-hidden="false"
                  class="bp3-collapse-body"
                  style="transform: translateY(-undefinedpx);"
                />
              </div>
            </div>
            <div
              class="Accordion--panel"
              data-open="false"
              data-testid="failureStrategy-panel"
            >
              <div
                class="Accordion--summary"
                data-testid="failureStrategy-summary"
              >
                <div
                  class="Accordion--label"
                >
                  pipeline.failureStrategies.title
                </div>
                <div
                  class="Accordion--chevron"
                />
              </div>
              <div
                class="bp3-collapse Accordion--collapse"
              >
                <div
                  aria-hidden="false"
                  class="bp3-collapse-body"
                  style="transform: translateY(-undefinedpx);"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
    </form>
  </div>
</div>
`;
